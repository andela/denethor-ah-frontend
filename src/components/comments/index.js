import CommentEntries from './CommentEntries';
import CreateComment from './CreateComment';
import Comment from './Comment';

export { CommentEntries, CreateComment, Comment };