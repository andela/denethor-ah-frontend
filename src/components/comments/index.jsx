import CommentEntries from './CommentEntries';

export { CommentEntries };