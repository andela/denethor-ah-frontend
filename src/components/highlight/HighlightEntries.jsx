import React from 'react';
import uuid from 'uuid/v1';
import PropTypes from 'prop-types';

import { HorizontalLine } from '../horizontalLine';
import './styles.scss';

const HighlightEntries = ({ highlights = [] }) => (
  <div>
    {highlights.length > 0 && <div><HorizontalLine />Highlighted Comments</div>}
    {
      highlights.map(({ comment, highlight }) => {
        return (
          <div key={uuid()} className='highlight-entry'>
            <div className='highlight-text'>
              <p>{highlight}</p>
            </div>
            <div className='highlight-comment'>
              <p>{comment}</p>
            </div>
          </div>
        );
      })
    }
  </div>
);

HighlightEntries.propTypes = {
  highlights: PropTypes.array
}

export default HighlightEntries;
