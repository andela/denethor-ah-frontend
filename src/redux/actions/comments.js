import actions from './types';


export const addComment = comment => (dispatch) => {
  dispatch({
    type: actions.ADD_COMMENT,
    comment
  });
};

export const removeComment = id => (dispatch) => {
  dispatch({
    type: actions.REMOVE_COMMENT,
    id
  });
};
