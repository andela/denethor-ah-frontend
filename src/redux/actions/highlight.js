import axios from '../../utils/axiosConfig';
import { ADD_HIGHLIGHT, GET_HIGHLIGHTS, CLEAR_HIGHLIGHTS } from './types';
import { escapeInputs } from '../../utils/escapeInputs';

const { API_ROOT_URL } = process.env;

export const addHighlight = (articleId, highlightDetails) => (dispatch) => {
  const escapedDetails = escapeInputs(highlightDetails);

  return axios
    .post(`${API_ROOT_URL}/articles/${articleId}/highlights`, escapedDetails)
    .then((response) => {
      const { data: { data: { highlight, message } } } = response;

      dispatch({
        type: ADD_HIGHLIGHT,
        payload: highlight
      });

      return { message, highlight };
    })
    .catch(({ response: { status, error } }) => {
      if ( status === 401) {
        return ({ error: 'Please login to access this feature' });
      }
      return ({ error });
    });
};

export const getHighlights = (articleId) => (dispatch) => {
  dispatch({ type: CLEAR_HIGHLIGHTS });

  return axios
    .get(`${API_ROOT_URL}/articles/${articleId}/highlights`)
    .then((response) => {
      const { data: { data: { highlights } } } = response;

      dispatch({
        type: GET_HIGHLIGHTS,
        payload: highlights
      });
    })
    .catch(() => { });
}