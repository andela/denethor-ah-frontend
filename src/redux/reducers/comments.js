import {
  ADD_COMMENT,
  REMOVE_COMMENT,
  ADD_COMMENT_FAILURE,
  GET_COMMENT_SUCCESS,
  COMMENT_IMPRESSION_SUCCESS,
  GET_COMMENT_LIKES_COUNT_SUCCESS
} from '../actions/types';

export const commentReducerDefaultState = {
  comments: [],
  loading: false,
  commentsLikes: []
};

const commentReducer = (state = commentReducerDefaultState, action) => {
  let {
    commentsLikes
  } = state;
  const {
    commentImpression
  } = action;

  switch (action.type) {
    case ADD_COMMENT:
      return {
        ...state,
        loading: true
      }

    case GET_COMMENT_SUCCESS:
      return {
        ...state,
        comments: [...action.comments],
        loading: false
      };

    case ADD_COMMENT_FAILURE:
      return {
        ...state,
        loading: false
      }

    case REMOVE_COMMENT:
      return state.filter(({
        id
      }) => id !== action.id);

    case COMMENT_IMPRESSION_SUCCESS:
      {
        const filteredCommentsLikes = commentsLikes.filter(commentLike => 
          commentLike.commentId !== commentImpression.commentId
        );

        return {
          ...state,
          commentsLikes: [...filteredCommentsLikes, commentImpression]
        }
      }

    case GET_COMMENT_LIKES_COUNT_SUCCESS:
      return {
        ...state,
        commentsLikes: [...action.commentsLikes]
      }

    default:
      return state;
  }
};

export default commentReducer;