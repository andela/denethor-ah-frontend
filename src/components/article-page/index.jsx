import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import RoundedImage from '../../components/RoundedImage';
import VerticalMargin from '../../components/VerticalMargin';
import HorizontalMargin from '../../components/HorizontalMargin';
import RatingStars from '../../components/RatingStars';
import RatingStarsBox from '../../components/RatingStarsBox';
import HorizontalLine from '../../components/horizontal-line';
import TextAreaInput from '../../components/TextAreaInput';
import CommentEntries from '../../components/CommentEntries';
import { getOneArticle, rateArticle, getArticleAvgRating, likeArticle, dislikeArticle, bookmarkArticle } from '../../redux/actions/articles';
import { toast } from 'react-toastify';
import './style.scss';
import { LikeNumberGroup , BookMarkNumberGroup, DisLikeNumberGroup } from '../like-number-group';
// import impressionsControl from '../../utils/impressionScroll';



export class SingleArticleView extends Component {

  async componentDidMount() {
    const { fetchArticle, getArticleAvgRating, match } = this.props;
    let articleId = match.params.articleId;
    await fetchArticle(articleId);
    getArticleAvgRating(articleId)
  }

  starClickHandle = (item) => {
    let articleId = this.props.match.params.articleId;
    this.props.rateArticle(item, articleId )
    .then(() => toast.success('Thank you for rating this article'))
    .catch(error => {
      if (error.response.data.message === 'You already rated this article'){
        return toast.error('You have already rated this article');
      } else if(error.response.data === 'Unauthorized'){
         toast.error('You need to login to rate an article');
        return  this.props.history.push('/login')
      }
      toast.error('Error while rating article');
    })
  }

  likeHandle = () => {
    let articleId = this.props.match.params.articleId;
    this.props.likeArticle(articleId)
    .then((response) => {
      const message = response.data.data.message;
      toast.success(`${message}`)
    })
    .catch(error => {
      if(error.response === 'Unauthorized'){
        return  toast.error('You need to login to like an article');
     }
       toast.error(`${error}`);
    })
  }

  dislikeHandle = () => {
    let articleId = this.props.match.params.articleId;
    this.props.dislikeArticle(articleId)
    .then((response) => {
      const message = response.data.data.message;
      toast.success(`${message}`)
    })
    .catch(error => {
      if(error.response === 'Unauthorized'){
        return toast.error('You need to login to dislike an article');
     }
      return toast.error('An error occurred while trying to dislike the article');
    })
  }

  bookmarkHandle = () => {
    let articleId = this.props.match.params.articleId;
    this.props.bookmarkArticle(articleId)
    .then(() => toast.success('Thank you for bookmarking this article'))
    .catch(error => {
      if(error.response === 'Unauthorized'){
        return toast.error('You need to login to bookmark an article');      
     } else if(error.response.status === 409){
       return toast.error('You have already bookmarked this article')
     }
      return toast.error('An error occurred while trying to bookmark this article');
    })
  }

  render() {
    const { match: { params: { articleId } }, articles = [] } = this.props;
    const article = articles.find(item => item.id === articleId) || {};
    const { 
      title, 
      featuredImage = '/assets/img/map-typing.jpg', 
      author: { username = '', imageUrl: userAvatarImage = '' } = {}, 
      body,
      averageRating,
      ratingsCount,
      comments = [],
    } = article;

    return (
      <div>
        <div className='banner-section'>
          <div className='banner-image'>
            <img src={featuredImage} alt='banner background' />
          </div>
          <div className='transparent-overlay' />
          <div className='banner-content'>
            <div className='auto-container'>
              <div className='banner-content-title'>{title}</div>
              <div className='avatar-text-group'>
                <div className='user-profile-picture'>
                  <RoundedImage imageSource={userAvatarImage} alt='profile picture'/>
                  <VerticalMargin className='hide-for-medium' size={50} />
                </div>
                <HorizontalMargin className='show-for-medium' size={10} />
                <div className='article-username-rate-group'>
                  <div className='article-header-username'>{username}</div>
                  <div className ='article-rating-container'>
                    <div className='article-header-rating'>
                      <RatingStars rateNumber={(averageRating)} />
                    </div>
                    <div className='article-header-rated-times'>
                      {!ratingsCount ? 'No ratings yet' : `rated by ${ratingsCount} user${ratingsCount > 1 ? 's' : ''}`}
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div className='body-section'>
          <div className='auto-container wide-padding'>
            <div className='article-body'>
              <div dangerouslySetInnerHTML={{ __html: body }} />
            </div>
            <HorizontalLine />
            <div className='section-article-info'>  
              <div className='rate-impression'>
                <p>Rate this Article</p> <RatingStarsBox starClickHandle= {this.starClickHandle.bind(this)} /> 
              </div>  
              <div className='impressions-container'>
                <div className='impression'>
                 <LikeNumberGroup likeCount = {0} onClick={this.likeHandle.bind(this)}/>
                </div>
                <div className='impression'>
                <BookMarkNumberGroup bookMarkCount = {16} onClick ={this.bookmarkHandle.bind(this)}/>
                </div>
                <div className='impression'>
                <DisLikeNumberGroup dislikeCount = {0} onClick ={this.dislikeHandle.bind(this)} />
                </div>
              </div>
              <VerticalMargin className='show-for-medium' size={10} />
              <div>
                <div className='section-comment-create'>
                  <div>Leave comments</div>
                  <TextAreaInput className='article-comment-field' text='Your message' />
                  <button></button>
                </div>
                <div className='section-user-comments'>
                  <CommentEntries comments={comments} />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

SingleArticleView.propTypes = {
  articles: PropTypes.array.isRequired,
  match: PropTypes.object.isRequired,
  fetchArticle: PropTypes.func.isRequired,
  rateArticle: PropTypes.func.isRequired,
  getArticleAvgRating: PropTypes.func.isRequired,
  history: PropTypes.object,
  likeArticle: PropTypes.func,
  dislikeArticle: PropTypes.func,
  bookmarkArticle: PropTypes.func,
}

const mapStateToProps = (state) => ({
  articles: state.articles
});

const mapDispatchToProps = (dispatch) => ({
  fetchArticle: (id) => dispatch(getOneArticle(id)),
  rateArticle: (rating, articleId) => dispatch(rateArticle(rating, articleId)),
  getArticleAvgRating: (articleId) => dispatch(getArticleAvgRating(articleId)),
  likeArticle: (articleId) => dispatch(likeArticle(articleId)),
  dislikeArticle: (articleId) => dispatch(dislikeArticle(articleId)),
  bookmarkArticle: (articleId) => dispatch(bookmarkArticle(articleId))
});

export default connect(mapStateToProps, mapDispatchToProps)(SingleArticleView);



