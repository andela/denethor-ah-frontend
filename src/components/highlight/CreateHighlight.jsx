import React, { Component } from 'react';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import PropTypes from 'prop-types';

import Comment from '../comments/Comment';
import { addHighlight } from '../../redux/actions/highlight';
import toastOptions from '../../utils/toastOptions';
import './styles.scss';

class CreateHighlight extends Component {
  state = {
    isOpen: true 
  }

  handleOnClick = async (comment) => {
    const { props: { isLoggedIn, articleId, highlightText, addHighlight } } = this;
    if (!isLoggedIn) return toast.error('Login to access this feature', toastOptions);

    const { error, message } = await addHighlight(articleId, { comment, highlight: highlightText });
    if (error) return toast.error(error, toastOptions);
    toast.success(message, toastOptions);
    setTimeout(() => this.setState({ isOpen: false }), 5000);
  }

  render() {
    const { props: { display, highlightText, id }, state: { isOpen }, handleOnClick } = this;

    return (
      <div className={`highlight-comment-textarea ${display && isOpen ? 'active' : ''}`}>
        <div className='ellipses'>{highlightText}</div>
        <Comment id={id} handleOnClick={handleOnClick} />
      </div>
    );
  }
}

CreateHighlight.propTypes = {
  readerId: PropTypes.string,
  articleId: PropTypes.string,
  highlightText: PropTypes.string,
  addHighlight: PropTypes.func,
  display: PropTypes.bool,
  id: PropTypes.string,
  text: PropTypes.string,
  isLoggedIn: PropTypes.bool
};

const mapStateToProps = ({ auth: { isLoggedIn = false } }) => ({ isLoggedIn });
const mapDispatchToProps = ({ addHighlight });

export default connect(mapStateToProps, mapDispatchToProps)(CreateHighlight);