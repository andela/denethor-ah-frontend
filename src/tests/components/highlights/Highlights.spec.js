import React from 'react';
import { mount } from 'enzyme';
import { Comment } from '../../../components/comments';
import { CreateHighlight } from '../../../components/highlight/CreateHighlight';
import HighlightEntries from '../../../components/highlight/HighlightEntries';

describe('Test for Highlights and Highlight Entries', () => {
  const highlights = [{
    comment: 'Blah Blah Blah',
    highlight: 'Blah Blah Blah'
  }]

  it('should render the comment entry section', () => {
    const wrapper = mount(<HighlightEntries highlights={highlights} />);
    expect(wrapper.length).toBe(1);
    expect(wrapper.find('.highlight-entry').length).toBe(1);
  });

  it('should call toast error when server error occurs', () => {
    const props = {
      readerId: 'sample-user',
      articleId: 'sample-article',
      highlightText: 'sample highlight',
      addHighlight: jest.fn(),
      display: true,
      id: 'sample-id',
      isLoggedIn: true
    };
    const wrapper = mount(<CreateHighlight {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Comment).length).toBe(1);
  });

  it('should update state as user is typing', () => {
    const handleOnClick = jest.fn();
    const wrapper = mount(<Comment handleOnClick={handleOnClick} />);
    wrapper.find('TextAreaInput').at(0).simulate('change', { target: { value: 'hi' } });
    wrapper.find('Button').at(0).simulate('click');
    expect(wrapper.state('comment')).toEqual('hi');
    expect(handleOnClick).toHaveBeenCalled();
  });
});