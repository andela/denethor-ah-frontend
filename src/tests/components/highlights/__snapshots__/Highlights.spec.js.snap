// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Highlights and Highlight Entries should call toast error when server error occurs 1`] = `
<CreateHighlight
  addHighlight={[MockFunction]}
  articleId="sample-article"
  display={true}
  highlightText="sample highlight"
  id="sample-id"
  isLoggedIn={true}
  readerId="sample-user"
>
  <div
    className="highlight-comment-textarea active"
  >
    <div
      className="ellipses"
    >
      sample highlight
    </div>
    <Comment
      handleOnClick={[Function]}
      id="sample-id"
    >
      <div
        className="comment-create-field"
      >
        <TextAreaInput
          className="article-comment-field sample-id"
          onChange={[Function]}
          placeHolder="Your message"
          placeholder="Your message"
          value=""
        >
          <textarea
            className="textarea-input article-comment-field sample-id"
            onChange={[Function]}
            placeholder="Your message"
            value=""
          />
        </TextAreaInput>
        <Button
          className="article-comment-field__button"
          isRequestSent={false}
          onClick={[Function]}
          type="submit"
          value="submit"
        >
          <button
            className="article-comment-field__button"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            submit
          </button>
        </Button>
      </div>
    </Comment>
  </div>
</CreateHighlight>
`;
